chrysom/client.go

```diff
@@ -129,140 +162,229 @@ func validateConfig(config *ClientConfig) error {
 	}
 	return nil
 }
+func shouldUseJWTAcquirer(options acquire.RemoteBearerTokenAcquirerOptions) bool {
+	return len(options.AuthURL) > 0 && options.Buffer != 0 && options.Timeout != 0
+}
 
-func determineTokenAcquirer(config ClientConfig) (acquire.Acquirer, error) {
-	defaultAcquirer := &acquire.DefaultAcquirer{}
-	if config.Auth.JWT.AuthURL != "" && config.Auth.JWT.Buffer != 0 && config.Auth.JWT.Timeout != 0 {
-		return acquire.NewRemoteBearerTokenAcquirer(config.Auth.JWT)
+func buildTokenAcquirer(auth *Auth) (acquire.Acquirer, error) {
+	if shouldUseJWTAcquirer(auth.JWT) {
+		return acquire.NewRemoteBearerTokenAcquirer(auth.JWT)
+	} else if len(auth.Basic) > 0 {
+		return acquire.NewFixedAuthAcquirer(auth.Basic)
 	}
+	return &acquire.DefaultAcquirer{}, nil
+}
 
-	if config.Auth.Basic != "" {
-		return acquire.NewFixedAuthAcquirer(config.Auth.Basic)
+func validateGetItemsInput(input *GetItemsInput) error {
+	if input == nil {
+		return ErrUndefinedInput
 	}
 
-	return defaultAcquirer, nil
+	if len(input.Bucket) < 1 {
+		return ErrBucketEmpty
+	}
+	return nil
 }
 
-func (c *Client) GetItems(owner string) ([]model.Item, error) {
-	request, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/store/%s", c.remoteStoreAddress, c.bucketName), nil)
+func (c *Client) makeRequest(owner, method, URL string, body io.Reader) (*http.Request, error) {
+	r, err := http.NewRequest(method, URL, body)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrNewRequestFailure, err.Error())
 	}
-	err = acquire.AddAuth(request, c.auth)
+	err = acquire.AddAuth(r, c.auth)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrAuthAcquirerFailure, err.Error())
+	}
+
+	if len(owner) > 0 {
+		r.Header.Set(store.ItemOwnerHeaderKey, owner)
+	}
+
+	return r, nil
+}
+
+func (c *Client) do(r *http.Request) (*doResponse, error) {
+	resp, err := c.client.Do(r)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrDoRequestFailure, err.Error())
 	}
+	defer resp.Body.Close()
 
-	request.Header.Set(store.ItemOwnerHeaderKey, owner)
+	var dResp = doResponse{
+		Code: resp.StatusCode,
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
```

Comment:
what happens if the response body is `nil`?

Reply:
It shouldn't be nil if the err was nil according to https://golang.org/pkg/net/http/#Client.Do